### ISSUE TITLE
add a way to return side-information from grad

### ISSUE THREAD

**Issue by mattjj** [google/jax#366]
For example, a user may want to return a statistic or auxiliary data structure from the Python function to which `grad` is applied, not to be considered as part of the value of the mathematical function to be differentiated.

**Comment by jekbradbury** [comment]
Following up from JAXers conversation: it's possible to do this by manually invoking `vjp` with more than one seed:
```python
import jax
def loss(x): return (x + x, x)
def mygrad(x):
  (l, extra), pullback = jax.vjp(loss, x)
  dx, = pullback((1.0, 0.0))
  return dx, extra
```
But as Dougal pointed out, even calling `jax.ad.vjp` with `jax.ad_util.zero` (rather than `0.0`) [doesn't avoid materializing](https://github.com/google/jax/blob/63608fb9715b33cfac64dc26548fcd7da35e97d3/jax/interpreters/ad.py#L79) the second seed (and performing unnecessary backward-pass computations).

So it would be nice to have an API that either accepts real symbolic zeros or allows selecting a single scalar output to receive a seed.

**Comment by j-towns** [comment]
@jekbradbury as you probably know, that's [the approach taken in Autograd](https://github.com/HIPS/autograd/blob/master/autograd/differential_operators.py#L142..L147).

Note that the extra flop cost will, I think, not be too bad for most use cases, because often `extra` is something that's computed during computation of `l` anyway (or doesn't take much to compute from some intermediate value during the computation of `l). This is the case in the trivial example you give. That means that the zeros will only be propagated through a small piece of graph before being merged into the gradient computation that's going on anyway. Does that make sense?

**Comment by mattjj** [comment]
Can you guys try out the code in #484, and let me know if you run into issues, or have a better way of organizing the API?
